src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs

```diff
@@ -66,31 +71,53 @@ public override Task<object> ReadRequestBodyAsync([NotNull] InputFormatterContex
 
                 var jsonSerializer = CreateJsonSerializer();
 
-                EventHandler<Newtonsoft.Json.Serialization.ErrorEventArgs> errorHandler = null;
-                errorHandler = (sender, e) =>
+                var successful = true;
+                EventHandler<Newtonsoft.Json.Serialization.ErrorEventArgs> errorHandler = (sender, e) =>
                 {
+                    successful = false;
+
                     var exception = e.ErrorContext.Error;
-                    context.ModelState.TryAddModelError(e.ErrorContext.Path, e.ErrorContext.Error);
+
+                    // Handle path combinations such as "" + "Property", "Parent" + "Property", or "Parent" + "[12]".
+                    var key = e.ErrorContext.Path;
+                    if (!string.IsNullOrEmpty(context.ModelName))
+                    {
+                        if (string.IsNullOrEmpty(e.ErrorContext.Path) || e.ErrorContext.Path[0] == '[')
+                        {
+                            key = context.ModelName + e.ErrorContext.Path;
```

Comment:
It's a little wierd to parse the `IsNullOrEmpty` case falling into here. I know it's a no-op, it just seems overly clever